light/beacon/binary_merkle.go

```diff
@@ -0,0 +1,555 @@
+	"errors"
+	"math/bits"
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/minio/sha256-simd"
+)
+
+// MerkleValue represents either a 32 byte value or hash node in a binary merkle tree/partial proof
+type (
+	MerkleValue  [32]byte
+	MerkleValues []MerkleValue
+)
+
+var (
+	MerkleValueT = reflect.TypeOf(MerkleValue{})
+	merkleZero   [64]MerkleValue
+)
+
+func init() {
+	hasher := sha256.New()
+	for i := 1; i < 64; i++ {
+		hasher.Reset()
+		hasher.Write(merkleZero[i-1][:])
+		hasher.Write(merkleZero[i-1][:])
+		hasher.Sum(merkleZero[i][:0])
+	}
+}
+
+// UnmarshalJSON parses a merkle value in hex syntax.
+func (m *MerkleValue) UnmarshalJSON(input []byte) error {
+	return hexutil.UnmarshalFixedJSON(MerkleValueT, input, m[:])
+}
+
+// VerifySingleProof verifies a Merkle proof branch for a single value in a
+// binary Merkle tree
+// Note: index is a generalized tree index; bottomLevel is optional (root being
+// level 0) and if greater than len(proof)-1 then the sibling subtrees closest
+// to the root are not stored and all-zero value subtrees are assumed (makes
+// proofs for 24 bit arrays more efficient).
+func VerifySingleProof(proof MerkleValues, index uint64, value MerkleValue, bottomLevel int) (common.Hash, bool) {
+	hasher := sha256.New()
+	var proofIndex int
+	for index > 1 {
+		var proofHash MerkleValue
+		if proofIndex < len(proof) {
+			proofHash = proof[proofIndex]
+		} else {
+			if i := bottomLevel - proofIndex - 1; i >= 0 {
+				proofHash = merkleZero[i]
+			} else {
+				return common.Hash{}, false
+			}
+		}
+		hasher.Reset()
+		if index&1 == 0 {
+			hasher.Write(value[:])
+			hasher.Write(proofHash[:])
+		} else {
+			hasher.Write(proofHash[:])
+			hasher.Write(value[:])
+		}
+		hasher.Sum(value[:0])
+		index /= 2
+		proofIndex++
+	}
+	if proofIndex < len(proof) {
+		return common.Hash{}, false
+	}
+	return common.Hash(value), true
+}
+
+// ProofFormat defines the shape of a partial proof and allows traversing a subset of a tree
+type ProofFormat interface {
+	children() (left, right ProofFormat) // either both or neither should be nil
+}
+
+// ProofReader allows traversing and reading a tree structure or a subset of it.
+// Note: the hash of each traversed node is always requested. If the internal
+// hash is not available then subtrees are always traversed (first left, then right).
+// If internal hash is available then subtrees are only traversed if needed by the writer.
+type ProofReader interface {
+	children() (left, right ProofReader) // subtrees accessible if not nil
+	readNode() (MerkleValue, bool)       // hash should be available if children are nil (leaf node), optional otherwise (internal node)
+}
+
+// ProofWriter allow collecting data for a partial proof while a subset of a tree is traversed.
+type ProofWriter interface {
+	children() (left, right ProofWriter) // all non-nil subtrees are traversed
```

Comment:
I think we should return an extra indicator if the proof is exhausted.
The nil children feels unintuitive.

Reply:
There is no such thing as "exhausted" here, this interface traverses a certain shape, passes data in the process but does not care about where that data comes from or where it goes. A `nil` child means that it does not want to traverse any deeper from that given node (hence no ProofWriter for its children). So I feel it is appropriate and intuitive to pass `nil` here.